### PR TITLE
Show whole symbol selector dialog (with Cancel and Help buttons) in some context

### PR DESCRIPTION
**Pull request by DelazJ**
Fixes [#17903](https://issues.qgis.org/issues/17903) and [#17902](https://issues.qgis.org/issues/17902)
In Layer Properties -> Symbology ->Categorized (or Graduated), clicking the [Change] button does not open the symbol selector dialog but the symbol selector widget, meaning that you miss Cancel and Help buttons. Also clicking a category symbol should not change the main button's.
Also fix an unreported crash that occurs when hitting ESC from a symbol selector widget (you may need two clicks)

### CODE REVIEW

File: src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp [hunk 1]
```diff
@@ -548,17 +548,39 @@ void QgsCategorizedSymbolRendererWidget::changeSelectedSymbols()
 
 void QgsCategorizedSymbolRendererWidget::changeCategorizedSymbol()
 {
+  QgsPanelWidget *panel = QgsPanelWidget::findParentPanel( this );
   QgsSymbol *newSymbol = mCategorizedSymbol->clone();
-  QgsSymbolSelectorWidget *dlg = new QgsSymbolSelectorWidget( newSymbol, mStyle, mLayer, nullptr );
-  dlg->setContext( mContext );
+  if ( panel && panel->dockMode() )
+  {
+    QgsSymbolSelectorWidget *dlg = new QgsSymbolSelectorWidget( newSymbol, mStyle, mLayer, panel );
```
Comment (elpaso) [inline src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp]: I'd suggest you to use `std::uniqe_ptr<QgsSymbolSelectorWidget>` here and `qgis::make_unique<QgsSymbolSelectorWidget>( newSymbol, mStyle, mLayer, panel );` (and remove the `delete`)

File: src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp [hunk 2]
```diff
@@ -958,12 +981,29 @@ void QgsGraduatedSymbolRendererWidget::changeSelectedSymbols()
 void QgsGraduatedSymbolRendererWidget::changeRangeSymbol( int rangeIdx )
 {
   QgsSymbol *newSymbol = mRenderer->ranges()[rangeIdx].symbol()->clone();
-  QgsSymbolSelectorWidget *dlg = new QgsSymbolSelectorWidget( newSymbol, mStyle, mLayer, nullptr );
-  dlg->setContext( mContext );
+  QgsPanelWidget *panel = QgsPanelWidget::findParentPanel( this );
+  if ( panel && panel->dockMode() )
+  {
+    QgsSymbolSelectorWidget *dlg = new QgsSymbolSelectorWidget( newSymbol, mStyle, mLayer, panel );
```
Comment (elpaso) [inline src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp]: Same as above.

File: src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp [hunk 3]
```diff
@@ -777,15 +782,14 @@ void QgsGraduatedSymbolRendererWidget::updateSymbolsFromWidget()
         if ( selectedIndexes.count() > 1 )
         {
           //if updating multiple ranges, retain the existing range colors
-          newRangeSymbol->setColor( mRenderer->ranges().at( rangeIdx ).symbol()->color() );
+          newRangeSymbol->setColor( mRenderer->ranges()[ rangeIdx ].symbol()->color() );
```
Comment (elpaso) [inline src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp]: Why did you change this? If I read the docs right, the [] operator returns a non const reference while at() does. You are not modifying there reference here so at() looks more appropriate to me.

**Review by elpaso — COMMENTED** [review]
LGTM, I just left a couple of notes.

File: src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp (hunk 1 above)
Comment (DelazJ) [reply src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp]: Hi @elpaso thanks for the review and sorry for the long silence. 
Sorry also to say that i do understand almost nothing in your message: my c++ knowledge is near zero and this PR is the fruit of some trial and errors (copy-paste) coding. I do not have time nor a build environment to give it another try so i'd really really appreciate any code sample i could just copy paste to comply with the review. Any helper?

File: src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp (hunk 3 above)
Comment (DelazJ) [reply src/gui/symbology/qgsgraduatedsymbolrendererwidget.cpp]: I will revert the change

**Comment by elpaso** [comment]
I pushed some minor fixes in https://github.com/qgis/QGIS/pull/6952
Closing this PR in favour of the new one.
